packages/url/src/is-url.native.js

```diff
@@ -1,4 +1,4 @@
-const URL_REGEXP = /((([A-Za-z]{3,9}:(?:\/\/)?)(?:[\-;:&=\+\$,\w]+@)?[A-Za-z0-9\.\-]+|(?:www\.|[\-;:&=\+\$,\w]+@)[A-Za-z0-9\.\-]+)(?::\d{2,5})?((?:\/[\+~%\/\.\w\-_]*)?\??(?:[\-\+=&;%@\.\w_]*)#?(?:[\.\!\/\\\w]*))?)/i;
+const URL_REGEXP = /((([A-Za-z]{3,9}:(?:\/\/?\/)?)(?:[\-;:&=\+\$,\w]+@)?[A-Za-z0-9\.\-]+|(?:www\.|[\-;:&=\+\$,\w]+@)[A-Za-z0-9\.\-]+)(?::\d{2,5})?((?:\/[\+~%\/\.\w\-_]*)?\??(?:[\-\+=&;%@\.\w_]*)#?(?:[\.\!\/\\\w]*))?)/i;
```

Comment:
In wondering whether the proposed revisions are in-fact technically valid, it eventually clicked with me that the extra slash is meant to represent the path name (e.g. `file:` is the protocol, then `//`, then `/foo` constituting a root-relative path). So in looking at how this is implemented, I wonder if this is something where we should be accounting for this as part of the pattern matching for the path, and not strictly as an optional extra slash following the `//`. The pattern is a bit unwieldy to untangle, so it's not immediately clear to me where this change would need to occur.

It's also a bit awkward that it's the second of three slashes which is made to be optional. Though I expect it works just the same.

```suggestion
const URL_REGEXP = /((([A-Za-z]{3,9}:(?:\/\/\/?)?)(?:[\-;:&=\+\$,\w]+@)?[A-Za-z0-9\.\-]+|(?:www\.|[\-;:&=\+\$,\w]+@)[A-Za-z0-9\.\-]+)(?::\d{2,5})?((?:\/[\+~%\/\.\w\-_]*)?\??(?:[\-\+=&;%@\.\w_]*)#?(?:[\.\!\/\\\w]*))?)/i;
```

Noting that the `//` is technically considered optional based on this pattern, I wonder if the following should be considered valid, as omitting the optional `//`, but still root-relative:

```
file:/foo
```

In the browser, this is considered valid, and evaluates the same as `file:///foo`:

```js
new URL( 'file:/ok' ).pathname
// "/ok"
```

Under this proposed implementation, I expect that would not be supported.

Reply:
Possibly related contrary information for the above-given example:

>- If url’s scheme is "file", then:
>   - If remaining does not start with "//", validation error.

https://url.spec.whatwg.org/#scheme-state

Reply:
>Noting that the // is technically considered optional based on this pattern, I wonder if the following should be considered valid, as omitting the optional //, but still root-relative:

At least based on relevant specification, it is optional:

>If c is U+002F (/), then set state to authority state.
>
>Otherwise, set state to path state, and decrease pointer by one.

https://url.spec.whatwg.org/#path-or-authority-state

The diagram here is a little easier to interpret, though technically not the specification we adhere to:

https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Generic_syntax

If the extra slash we're accounting for is relevant for the path, then the pattern should match it separately from the optional authority `//`.

Reply:
@aduth I did some changes to support the case you mention, have a look to see if you agree.

Reply:
@SergioEstevao I'm not seeing any changes. Did you push them? GitHub was having an outage incident this morning, so possible that it was missed?

Reply:
My previous comments were admittedly a bit of a winding brain dump. In short, it's not clear to me if we should consider `file:/foo` as "valid". It's tolerated by the `URL` constructor and thus would return `true` in the browser, so there's an argument in favor of consistency. But as I note in https://github.com/WordPress/gutenberg/pull/20435#discussion_r384706295, the specification is fairly clear that a double-slash should be required to be considered valid.

Reply:
Did you come up with this regular expression or find it somewhere? I was trying to decipher it yesterday, and it felt to me like there's some redundancies around `www.`, and user/pass matching. It made it hard for me to figure out how best to change it, but ultimately it felt like the problem is in-part because there's an expectation that one or more characters exist before the pathname can start (which is what the changes here are proposing is an incorrect assumption).

Notably, the occurrences of: `[A-Za-z0-9\.\-]+`

Reply:
I've updated the expression that @etoledom use. I think he used the one for Javascript defined here:  https://urlregex.com/

I'm tempted to replace it by the one defined for HTML 5:
`^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?`

Reply:
> I'm tempted to replace it by the one defined for HTML 5:
^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?

I've tried this and it's no good, I think the current version of the commit is the best one that matches all the situations.

Reply:
@aduth ^